src/CssToInlineStyles.php

```diff
@@ -1,682 +1,242 @@
-            $chunks[1] = trim($chunks[1]);
-
-            // add to pairs array
-            if (!isset($pairs[$chunks[0]]) ||
-                !in_array($chunks[1], $pairs[$chunks[0]])
-            ) {
-                $pairs[$chunks[0]][] = $chunks[1];
+            foreach ($elements as $element) {
+                $this->calculatePropertiesToBeApplied($element, $rule->getProperties());
             }
         }
 
-        // sort the pairs
-        ksort($pairs);
-
-        // return
-        return $pairs;
-    }
-
-    /**
-     * Should the IDs and classes be removed?
-     *
-     * @return void
-     * @param  bool [optional] $on Should we enable cleanup?
-     */
-    public function setCleanup($on = true)
-    {
-        $this->cleanup = (bool) $on;
-    }
-
-    /**
-     * Set CSS to use
-     *
-     * @return void
-     * @param  string $css The CSS to use.
-     */
-    public function setCSS($css)
-    {
-        $this->css = (string) $css;
-    }
+        $elements = $xPath->query('//*[@data-css-to-inline-styles]');
 
-    /**
-     * Set the encoding to use with the DOMDocument
-     *
-     * @return void
-     * @param  string $encoding The encoding to use.
-     *
-     * @deprecated Doesn't have any effect
-     */
-    public function setEncoding($encoding)
-    {
-        $this->encoding = (string) $encoding;
-    }
+        foreach ($elements as $element) {
+            $propertiesToBeApplied = $element->attributes->getNamedItem('data-css-to-inline-styles');
+            $element->removeAttribute('data-css-to-inline-styles');
 
-    /**
-     * Set HTML to process
-     *
-     * @return void
-     * @param  string $html The HTML to process.
-     */
-    public function setHTML($html)
-    {
-        $this->html = (string) $html;
-    }
+            if ($propertiesToBeApplied !== null) {
+                $properties = unserialize(base64_decode($propertiesToBeApplied->value));
+                $this->inlineCssOnElement($element, $properties);
+            }
+        }
 
-    /**
-     * Set use of inline styles block
-     * If this is enabled the class will use the style-block in the HTML.
-     *
-     * @return void
-     * @param  bool [optional] $on Should we process inline styles?
-     */
-    public function setUseInlineStylesBlock($on = true)
-    {
-        $this->useInlineStylesBlock = (bool) $on;
+        return $document;
     }
 
     /**
-     * Set strip original style tags
-     * If this is enabled the class will remove all style tags in the HTML.
+     * Store the calculated values in a temporary data-attribute
      *
-     * @return void
-     * @param  bool [optional] $on Should we process inline styles?
+     * @param \DOMElement $element
+     * @param Property[]  $properties
+     * @return \DOMElement
      */
-    public function setStripOriginalStyleTags($on = true)
-    {
-        $this->stripOriginalStyleTags = (bool) $on;
-    }
+    protected function calculatePropertiesToBeApplied(
```

Comment:
does this need to be an extension point ? If no, you may want to make it private